src/BaGet.Hosting/FileProvider/BaGetUIFileProvider.cs

```diff
@@ -0,0 +1,136 @@
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.FileProviders.Physical;
+using Microsoft.Extensions.FileSystemGlobbing;
+using Microsoft.Extensions.Primitives;
+
+namespace BaGet.Hosting
+{
+    public class BaGetUIFileProvider : IFileProvider
+    {
+        private readonly DirectoryInfo _root;
+        private readonly string _pathBase;
+        private readonly PhysicalFilesWatcher _watcher;
+        private readonly Matcher _matcher;
+        private readonly ConcurrentDictionary<string, CachedFileInfo> _fileCache;
+
+        public BaGetUIFileProvider(DirectoryInfo root, string pathBase)
+        {
+            _root = root;
+            _pathBase = pathBase;
+            _watcher = new PhysicalFilesWatcher(root.FullName, new FileSystemWatcher(root.FullName), true);
+            _matcher = new Matcher().AddInclude("**/index.html").AddInclude("**/*.js");
+            _fileCache = new ConcurrentDictionary<string, CachedFileInfo>();
+        }
+
+        public IDirectoryContents GetDirectoryContents(string subpath)
+        {
+            return new DirectoryContents(this, new DirectoryInfo(Path.Combine(_root.FullName, subpath.TrimStart('/'))));
+        }
+
+        public IFileInfo GetFileInfo(string subpath)
```

Comment:
I'd prefer it if this feature was a little more "pay for play". Right now we always cache files in memory even the `PathBase` configuration wasn't set. It'd be better if we don't cache or replace `__BAGET_PLACEHOLDER_...` if `PathBase` isn't configured.

Reply:
I know three options to implement a custom pathbase with ASP.NET and Node:
1. Replace `__BAGET_PLACEHOLDER_...` in every affected file and cache them. I would not expect any performance penalty as files are kept in memory. An instance of BaGet occupies ~100 MB RAM and caching all JavaScript files would another ~700 KB.
2. Replace `__BAGET_PLACEHOLDER_...` only in `index.html` and implement code in `config.tsx` to dynamically replace `__BAGET_PLACEHOLDER_...` with an empty string. This would reduce the additional memory usage for standard deployments to less than 3 KB.
3. Omit placeholders in `index.html` and use the concept from 2. for JavaScript. If a pathbase is configured we could edit `index.html` with regular expressions. This would require a full test coverage because we would have to rely on implementation details of the Node compiler.

If you know a better way to implement a custom pathbase please describe it to me or tell me which of these three options would like to have in BaGet.

Reply:
@loic-sharma What's your opinion on that?